Updates to the release strategy

mattcaswell opened the pull request:
Remove out of date sections, and add some words about stability.

In WIP until discussion has taken place.

mattcaswell committed:
Add a stability policy to the release strategy

t-j-h commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,20 @@ <h5>
-	    <li>19th June 2018, beta release 6 (pre8)</li>
-	    <li>Release readiness check following pre8 release (new release
-	        cycles added if required)</li>
+	    <li>No existing public interface can be removed or modified until its
```
We need to define what we mean by "public interface" here clearly as there have been rather different views on that.
The two views I have seen regularly expressed:
1) in the header files that are installed
2) in the pod documentation files that are installed

Clearly 2) is a much smaller list than 1).

mattcaswell replied:
In my mind it is clearly (1)

tmshort replied:
It's actually a problem that (2) is smaller than (1). Lots of things still need documentation.

richsalz replied:
FWIW, I remember trying to remove something because it wasn't documented and Tim was the leading proponent of "it's in the header file" but that most of the team agreed and I was definitely in the rough. 
```
; ./util/find-doc-nits -u
# Duplicate i2d_PublicKey in doc/man3/d2i_X509.pod and doc/man3/d2i_PrivateKey.pod
# Found 4577 in util/libcrypto.num
# Found 1607 missing from util/libcrypto.num

# Found 499 in util/libssl.num
# Found 57 missing from util/libssl.num

# Checking macros (approximate)
# Found 229 macros missing (not all should be documented)
; 
```

richsalz replied:
Try the -d flag also.

t-j-h replied:
So working those rules taking the function DES_cbc_encrypt as an example - it is in the header file des.h and it is not currently marked deprecated. Do we consider the EVP-level API a replacement?
And in order to remove the API we have to mark it as deprecated first in a release and then wait at least 5 more years before it is removed which can only happen in a API-breaking release itself?

mattcaswell replied:
Yes. Although we could (potentially) move it to a "libcrypto-legacy" library. The structure could be made opaque (according to the rules stated for making things opaque). And the internal implementation could be changed to call EVP.

kaduk replied:
Functions that have always been documented as being not guaranteed to be API-stable would presumably get an exception (or, if later documented that way, after the same timeout as for removal).

Some other projects I work with have the luxury of an explicit export symbol list to be clear about what symbols are in the public API (intersected with the public headers, since in some cases internal symbols from one library are exposed to other libraries in the same implementation but not to the public).  Our export list is generated from the headers at present, so that just leaves us with the public headers alone, even if that leaves us in an awkward position regarding some APIs that really ought not to have existed.

mattcaswell replied:
I put some text in to define public interface as "any function or macro declared in a public header file"

tmshort commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,20 @@ <h5>
-	        cycles added if required)</li>
+	    <li>No existing public interface can be removed or modified until its
+	    replacement has been in place in an LTS stable release and the original
+	    function has been documented as deprecated for at least 5 years.</li>
```
So, for an API to be *modified*, it would have to be documented as deprecated? Not sure that makes sense. The most recent example of a possible API change is to `SSL_set_options()`, where it currently takes and returns a `long`. Changing that to a `uint64_t` allows for a consistent number (64) of options, where as right now, it depends on the compiler defintion of `long`. Obviously, changing the type breaks the ABI, but there's a subtle difference between changing an input parameter to 64bits, and the return value to 64bits. The first could be considered API (source) compatible, since the compiler will automatically extend the input to match function signature. But that doesn't happen for the return value.

mattcaswell replied:
That's a good point. I was thinking of modifications more along the lines of adding or removing parameters. I'm not sure about changing types.

If you changed SSL_set_options() to take `uint64_t` (which is probably API compatible) - wouldn't that necessarily imply also changing SSL_get_options() to return a `uint64_t` which would be API breaking.
I'd be interested in @vdukhovni's thoughts.

mattcaswell replied:
I suppose there is also the example of "const"ifying parameters - which we've generally allowed.

vdukhovni replied:
The discussion here is about the API, not ABI, and the example of widening an integer parameter does not introduce an API break. Existing calls that pass a long continue to work when the target type changes to uint64_t, through the magic of C prototypes and integral promotion rules. Similarly, constification of non-pointer input parameters is harmless.

The goal here is to avoid abrupt API transitions where we have one API in OpenSSL version X and a different incompatible API in version X+1, with no common way to perform some functions across the two versions. The notion of X and X+1 is specified as a time range, since our downstream consumers don't necessarily adopt each and every version, but rather do batch upgrades at major plaform releases, ...  We can debate whether 5 years is "just right", or some other number is better, but the idea should be to provide plenty of opportunity for users doing upgrades to have a real migration path where their software compiles against both the old and the new, and then get to gradually fix the deprecation warnings before the next major upgrade.  That way an OS distribution can rebuild all its packages, and work on modernization of the API without having upgrading OpenSSL blocked on porting all the applications.

kaduk replied:
I agree with Viktor -- we definitely need to provide the compat APIs in the previous version to allow for a smooth transition. In particular, I think the following text about adding opaquification accessors should also allow adding compat APIs.
(It doesn't sound like we need to debate the actual timeout right now, but I would personally go for more like 2-3 years than 5.)

mattcaswell replied:
I clarified "modified" to say "or modified (in an API incompatible manner)"

vdukhovni commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,20 @@ <h5>
+	    <li>When structures are made opaque, any newly required accessor
+	    macros are added in a feature release of the extant LTS release
+	    and all supported intermediate successor releases (not in security-fix
+	    only mode). Non-trivial accessors may be static inline functions, that
```
I am bit concerned about the "(security fix only mode)" potentially becoming a blocker for upgrades, if some major OS distribution bundles a non-LTS release that goes "(security fix only)" and now needs to support applications running against that, and the release that wants to add the opaque accessors.  We may face this problem with OpenSSL 1.1.0 for example. So I think that adding conditionally enabled macros to header files to support newly required accessors should sometimes also apply to releases that otherwise only get security fixes.

This is in part why my original suggestion explicitly mentioned conditional processing of the new macros based on a compiler flag that indicates that the code is being forward ported to a later API.  An implementation detail, but one that allows us to more leeway to make changes otherwise proscribed for stability reasons.

levitte replied:
Side question: how do you plan to have macros and static inline functions work against an opaque structure?

vdukhovni replied:
The text is about making the new accessors available in the **previous** release, where the structures are not yet opaque, and had no accessors.  That is users who're migrating to the new accessors should also find find them in the extant release, rather than everyone having to roll their own.

mattcaswell replied:
I removed "security fix only mode" from that text.

tmshort replied:
Internally, we've run into instances where we've back-ported accessors and it caused problems with software packages that implemented their own accessors.

kaduk replied:
This would presumably be less of a problem if the upstream accessors were introduced in all our branches simultaneously, saving the other software from needing to ever implement their own.

t8m replied:
I think if this is clearly stated and followed as policy that accessors are always backported to all live branches, the 3rd party software will handle it fine. The worst thing is to keep the policy undefined - i.e. sometimes backport it and sometimes not or after a long delay, then the 3rd party implementation might really clash.

mspncp replied:
The refusal to provide an officially supported and maintained compatibility header (e.g. `<openssl/compat.h>`) to ease the transition from direct member access to accessor functions was the most significant obstacle for upgrading from 1.0.2 to 1.1.0 IMHO. It is a pity that every downstream project had to reinvent the wheel. (I say this as somebody who is currently porting our company's code from 1.0.2 to 1.1.0...)

mlindner commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,20 @@ <h5>
+	    and all supported intermediate successor releases (not in security-fix
+	    only mode). Non-trivial accessors may be static inline functions, that
+	    compile to new code in the calling application.</li>
+	    <li>No cosmetic function renaming for consistency.</li>
```
Just an openssl user here but renaming functions for naming consistency is something that has been long overdue in OpenSSL and I'm glad it is occurring. We shouldn't stop that from continuing.

vdukhovni replied:
If all that's happening is a name change, like the changes from `..._create()` to `..._new()` for some functions between 1.0.2 and 1.1.0, then the disruption caused to existing software dwarfs any benefit from the renaming. We failed to provide a smooth transition to the new names, and though the old names did not fit into the new naming pattern, they did the job.  Such renaming would have to be handled with care (the above 5 year deprecation timeline, backport of new names to old releases, ...) and the pain would have to be worth the effort.  Perhaps sometimes this will be worth the trouble, but often the old names are good enough.

kaduk replied:
In light of the 5-year timeline, this would not as much be "replacing" as "adding new APIs that match the naming scheme, but providing a shim layer so the old APIs still work".

sam-github replied:
Renaming functions for consistency hurts all existing users of the existing name, but it helps all the future users coming to OpenSSL, and also helps the OpenSSL maintainers if it can cut down the number of questions and bug reports by confused users in the long run. I think modernizing the API is well worth doing, as long as the transition is smooth, slow and ordered. The proposed 5 year deprecation cycle seems reasonable to me.

kaduk replied:
I think we seem to all be saying that a blanket prohibition on cosmetic renaming is overkill, and that it can be safe to do with sufficient care?

mlindner commented:
The company I work for uses OpenSSL heavily and we're still on 1.0.2 and looking forward to moving to 1.1.1. It's going to force a lot of code rewriting but this will be cleaner code than we had anyway and cause less bugs.

I think the thing to be looking at when making breaking changes is making sure the OpenSSL team answers the question "Are all possible ways this function could be used have a drop in replacement alternative method of performing them? Ideally with little rewriting of code." If that is handled properly then there shouldn't be as much an issue with making breaking changes. Also any breaking changes that are caused should ideally cause compile time errors/warnings rather than hidden subtle breakages. Renaming functions/creating new ones is better than modifying the functionality of existing ones. In fact I'd prefer a strict "never modify functionality of documented LTS release functions, only remove/deprecate them".

vdukhovni commented:
> @mlindner The company I work for uses OpenSSL heavily and we're still on 1.0.2 and looking forward to moving to 1.1.1. It's going to force a lot of code rewriting but this will be cleaner code than we had anyway and cause less bugs.

Sure, API changes may be par for the course for a single consumer of the library willing to port a handful of products.  The picture is dramatically different for OS distributions with thousands of dependent packages that need to be ported.  Since we want to see reasonably rapid uptake of the latest greatest version of OpenSSL, even as we evolve the API, we need to provide much better migration paths than we did from 1.0.2 to 1.1.0.  Yes, there are many improvements in 1.1.0, and yes we want users to port to the new way of doing things, but the transitions need to be much less painful, and it should be possible to "just recompile" old code and get it deployed with the new library, and **then** gradually port to the new way of writing the code.  One can test the need for such rewrites by building against a build Configured with "no-deprecated" or definiing `OPENSSL_API_COMPAT` to the latest version number.

levitte committed:
Release strategy: add text on the 3.0.0 versioning scheme

levitte committed:
Remove the 1.1.1 time table and add support information

levitte committed:
Generalise the descriptions of alpha, beta, and release criteria

mspncp commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,20 @@ <h5>
-          amended at any time as the need arises.</p>
-
+	  <p>The following stability rules apply to all changes made to code
+	  targetted for a version after 1.1.1:</p>
```
typo: targeted

mattcaswell replied:
Fixed

mattcaswell commented:
New updates pushed, addressing comments so far

levitte commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,22 @@ <h5>
+	    API incompatible manner) until its replacement has been in place in an
+	    LTS stable release and the original function has been documented as
+	    deprecated for at least 5 years. A public interface is any function or
+	    macro declared in a public header file.</li>
```
The language here is confusing.

If a function is modified, it means it's name remains the same but it's signature changes. 
If it's replaced, it means a new symbol is added, and the old symbol is deprecated and removed later.

Those are two different ways of doing things, and this text seems to mix them up.

tmshort replied:
Changing types to functions, specifically `int` to `ssize_t` or `size_t` is changing an interface, but is usually a source-compatible. Or even the length of the options field/API from `long` to `uin64_t`. These types of changes ("size_t'ifying") should be allowed at ABI-breaking boundaries without these restrictions.

kaduk commented on policies/releasestrat.html:
```diff
@@ -55,65 +56,22 @@ <h5>
-	  <ul>
-	    <li>Feature complete/Feature freeze (but may include a non-final implementation of the TLSv1.3 specification)</li>
-	    <li>Bug fixes only</li>
-	  </ul>
```
I guess there's not enough here to make it a clear win to keep the alpha/beta definitions in place until the next major release.

levitte replied:
I actually disagree with the removal, and rather think it should be present, but more generalised.  See #104

mattcaswell replied:
> I actually disagree with the removal, and rather think it should be present, but more generalised. See #104 

I've added the commits from #104 here and kept the more general version of this text.

mspncp commented:
ping?

mattcaswell commented:
I updated this to take account of the latest feedback. Please take another look.

mattcaswell commented:
I merged the updates proposed in #104 into this PR so we can see the full picture.

levitte commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
+	    are unlikely to break source compatibility.
+	    <li>No existing public interface can be removed until its replacement
+	    has been in place in an LTS stable release and the original function has
+	    been documented as deprecated for at least 5 years. A public interface
```
The mixture of "public interface" and "original function" disturbs me, and "original function" feels linguistically clumsy.  May I suggest the following reprase?

``` html
	    <li>No existing public interface can be removed until it has been
            documented as deprecated for at least 5 years and its replacement
            has been in place in an LTS stable release. A public interface
```

kaduk replied:
I'm not super-picky about which wording, but I wonder if we could get away with 3 or even 2 years.

mattcaswell replied:
2 or 3 years sounds quite short to me.

levitte replied:
Regardless of length of time (I'm fine with 5 years), I'll still pick at the phrasing :wink:

levitte replied:
I don't mind shortening the time...  but oh, this comes back to phrasing again!  I assumed that the 5 years only referred to deprecation, is that correct?  With the current phrasing, "at least 5 years" could be understood to refer both to the deprecation and the presence in an LTS stable release.

Regardless of phrasing, I'd apply much more importance in the replacement's presence in a LTS release than the length of deprecation.

mattcaswell replied:
I updated the phrasing to try and make this clearer.

levitte commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
+	    is any function or macro declared in a public header file.</li>
+	    <li>When structures are made opaque, any newly required accessor
+	    macros or functions are added in a feature release of the extant LTS
+	    release and all supported intermediate successor releases.</li>
```
I'm a little confused by this...  does that mean that  will nominate MAJOR releases as LTS and include all MINOR releases therein?

mattcaswell replied:
An example: let's say we have releases 3.0.x and 3.1.x currently in support and 3.0.x is nominated as LTS. During the development of 4.0.0 we make a structure opaque. This means 3.0.x will have the accessor macros/functions backported to it (because it is an LTS release). They will also be backported to 3.1.x (since it is an intermediate release).

levitte replied:
I guess it's "feature release of the ... feature release" that throws me off. The way I have understood things is that feature releases have a new *x* in 3.*x*.0, and talking about a feature release of something that has an unchanging *x* makes that confusing...

sam-github replied:
At the beginning of  the document, you define MAJOR (no changes within a MAJOR can break API/ABI). But you never defined MINOR.PATCH, though you do say you aren't adopting all of semver, but that is only a statement on what they are not, not what they are. I think they should be defined.

> 3.0.x will have the accessor macros/functions backported to it

This would of course be illegal under semver. Introduction of new features requires a MINOR bump, so only changes that introduce no new features (essentially, bug fixes) can be backported to a MAJOR.MINOR.

What I _think_ you are intending after reading this paragraph and the one on top about MAJOR is this:

MAJOR.STABLE.SERIAL 

- MAJOR: API/ABI incompatible changes bump this
- STABLE: identifies a release line that shall only receive "stable" changes. occaisonally (when?) stable lines will be designated LTS. Both new features and bug fixes are allowed in a stable line (if they are considered stable)
- SERIAL: consecutive releases in a stable line get a new SERIAL number

Is something like this the intention?

You might want to avoid MINOR.PATCH, maybe its my perspective from node/npm which adheres very strongly to semver, but even using those words implies semver to me. But maybe that's not true in other software communities.

t8m replied:
> An example: let's say we have releases 3.0.x and 3.1.x currently in support and 3.0.x is nominated as LTS. During the development of 3.2.0 we make a structure opaque. This means 3.0.x will have the accessor macros/functions backported to it (because it is an LTS release). They will also be backported to 3.1.x (since it is an intermediate release).

I am confused now. How could you make something opaque in minor release 3.2.0 when this would break API/ABI from 3.0.0?

mattcaswell replied:
> I am confused now. How could you make something opaque in minor release 3.2.0 when this would break API/ABI from 3.0.0?

Oh. Sorry for the confusion. That's because my example is bad. Let's make the example this instead (I also updated the text above accordingly to save confusing anyone else):

An example: let's say we have releases 3.0.x and 3.1.x currently in support and 3.0.x is nominated as LTS. During the development of 4.0.0 we make a structure opaque. This means 3.0.x will have the accessor macros/functions backported to it (because it is an LTS release). They will also be backported to 3.1.x (since it is an intermediate release).

mattcaswell replied:
> MAJOR: API/ABI incompatible changes bump this
> STABLE: identifies a release line that shall only receive "stable" changes. occaisonally (when?) stable lines will be designated LTS. Both new features and bug fixes are allowed in a stable line (if they are considered stable)
> SERIAL: consecutive releases in a stable line get a new SERIAL number

Essentially, yes, although I think we will be sticking with MAJOR.MINOR.PATCH since this is already baked into the code. It could of course be changed but it's a more significant change. I think "minor" and "patch" have a meaning outside of semver.

Mostly we only allow bug fixes in a minor release - but there are exceptions. Notably for backport of accessors, and possibly "by exception" for other things.

sam-github replied:
Keeping minor/patch is reasonable, and the overall scheme is, too. I do think text somewhat like the above clearly stating what minor.patch means is missing from the document.

levitte replied:
I see no reason to go away from MAJOR.MINOR.PATCH.  Those are words for the three positions of a version number that have existed in that form for a looooong time (my memory on this takes me back to the 80's, although as MAJOR.MINOR-PATCH).  Semver is a much later effort to formalize and streamline this form of version number, but that doesn't mean that semver "owns" this version number form or the words used to describe it.

mattcaswell replied:
I added a description of each of MAJOR, MINOR and PATCH.

sam-github commented on policies/releasestrat.html:
```diff
@@ -20,27 +20,54 @@ <h5>
+	  <p>
+	    With this format, API/ABI compatibility will only be guaranteed
+	    for the same MAJOR version number. Previously we guaranteed
+	    API/ABI compatibility across the same MAJOR.MINOR combination.
```
I don't understand this. Its the "only" placement that confuses me, it seems to be misplaced. You are making things more strict, so shouldn't this be:

> With this format, API/ABI compatibility will be guaranteed for all releases with the same MAJOR version number. Previously we guaranteed API/ABI compatibility only across the same MAJOR.MINOR combination.

In other words, it used to be allowed to break API from 4.5.x to 4.6.x, but now, API can be broken only from 4.x.y to 5.x.y?

mspncp replied:
@sam-github I think you misinterpreted "only" in the sense of comparing the curren to the previous scheme (i.e, comparing the first sentence with the second). But the "only" in the first sentence referred  to the "same MAJOR version" of the same sentence.

My suggestion: either omit "only" from the first sentence (or add "only" to the second, as well).

sam-github replied:
I think omitting only from the first sentence, or my suggested text, both work. "Only" everywhere is still open for misinterpretation.

mattcaswell replied:
I dropped the word "only" here.

sam-github commented on policies/releasestrat.html:
```diff
@@ -20,27 +20,54 @@ <h5>
+	    This more closely aligns with the expectations of users who are
+	    familiar with semantic versioning.  However, we have not adopted
+	    semantic versioning in the strict sense of its rules, because it
+	    would mean changing our current LTS policies and practices.
```
Perhaps this isn't the right place, but what about semver is a problem for your current LTS policies and practices? Could you point me to another conversation?

mspncp replied:
See for example https://github.com/openssl/openssl/issues/5042#issuecomment-444204003 and ff.

sam-github replied:
I saw a statement of "we aren't going to do it the semver way", but no rationale for why. Saying because its current policy is a head-scratcher, because the policy is changing: API breaks are no longer allowed in MINOR. Not allowing API additions in PATCH could occur at the same time, and that would be all the way to semver. LTS policies are not covered by semver, of course, those are project specific.

Node has LTS releases and uses semver. Within a LTS-designated major, if we decide to add accessors (or something of the like), we bump the minor. No big deal. Not all new features go back into LTS. They need to be considered stable, and exist in a current release for some time (for bug shakedown). Eventually LTS goes into "maintenance mode", where we don't do anything except critical bug fixes. I don't think node's stability requirements are fundamentally that different from OpenSSL, other than that your LTS are much longer (a good thing, thank you). What we do is described here, in case its of interest: https://github.com/nodejs/Release#release-plan

Anyhow, I'm done kibitzing. The policy described here will work for us as OpenSSL consumers.

mspncp commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
-	    <li>No open Coverity issues (not flagged as "False Positive" or "Ignore")</li>
-	    <li>TLSv1.3 RFC published (with at least one beta release after the publicaction)</li>
+	    <li>
+	      All open github issues/PRs older than 2 weeks at the time of
```
Just a suggestion (add a "need" and get rid of the double "to be")
```
All open github issues/PRs older than 2 weeks at the time of
release need to be assessed for relevance to the version being released. 
```

mattcaswell replied:
Done

mspncp commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
+	      All open github issues/PRs older than 2 weeks at the time of
+	      release to be assessed for relevance to the version to be
+	      released. Any flagged with the a milestone for the version
+	      to be released must be closed (see below)
```
Add trailing period.

mattcaswell replied:
Done

mspncp commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
+	      Clean builds in Travis and Appveyor for two days
+	    </li>
+	    <li>
+	      run-checker.sh to be showing as clean 2 days before release
```
run-checker.sh succeeds on 2 consecutive days before release

_(edited: succeds -> succeeds)_

mspncp commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
+	    </li>
+	    <li>
+	      No open Coverity issues (not flagged as "False Positive" or
+	      "Ignore")
```
Side Note: then let's hope that Coverity will not be down again ;-) It is back online now, but still in readonly mode:

> Coverity Scan is back up and running in read-only mode. This means you can access your previous scan projects. We have re-built Scan using a cloud platform that will reduce dependencies and hopefully improve overall performance. We expect to be ready for new projects within the next few days once the final migration work is complete. We thank you for your patience as we work to make the platform better. For any questions in the mean time, please email sig-scan-admin@synopsys.com or visit community.synopsys.com.

mspncp commented on policies/releasestrat.html:
```diff
@@ -55,65 +82,92 @@ <h5>
+	      Working as intended
+	    </li>
+	    <li>
+	      Deliberate decision not to fix until a later release (this
```
...not to fix _the issue_ until...

mattcaswell replied:
Done

mattcaswell committed:
fixup! Add a stability policy to the release strategy

mattcaswell commented:
Fixup commit pushed addressing all comments received so far with the exception of @kaduk's suggestion of a shorter time period (2 or 3 years) for the deprecation timeout than the 5 years that this text currently says. I've not seen any discussion on that point.

levitte commented on policies/releasestrat.html:
```diff
@@ -142,7 +149,7 @@ <h5>
 	    </li>
 	    <li>
-	      Deliberate decision not to fix until a later release (this
+	      Deliberate decision not to fix thjis issue until a later release (this
```
This looks like a vi artifact :wink:

mattcaswell replied:
Fixed.

mattcaswell committed:
fixup! Add a stability policy to the release strategy

levitte commented on policies/releasestrat.html:
```diff
@@ -167,9 +167,9 @@ <h5>
-	    is any function or macro declared in a public header file.</li>
+	    has been in place in an LTS stable release. The original function must
+	    also have been documented as deprecated for at least 5 years. A public
+	    interface is any function or macro declared in a public header file.</li>
```
I think we need to say "structure" as well.

levitte replied:
Oh, and just wanted to say that I like the rephrase you ended up doing.  Thank you.

mattcaswell replied:
I added "structure" as well. I also changed "The original function" to say "The original interface".

t-j-h commented on policies/releasestrat.html:
```diff
@@ -20,27 +20,61 @@ <h5>
+	      to better meet developers' and vendors' expectations. Letter
+	      releases, such as 1.0.2a, exclusively contain bug and security
+	      fixes and no new features. Minor releases that change the
+	      last digit, e.g. 1.1.0 vs. 1.1.1, can and are likely to
```
This still remains with the problematic wording - where we labelled major.minor.fix.patch as the actual scheme and talked about "major" and "minor" releases in this separate context. That different usage should be removed in this description so we removed the opportunity for that confusion being repeated.

The simple fix is to change "Minor releases" to "Releases" and the problem goes away in this wording.

mattcaswell replied:
I applied the suggested fix

mattcaswell committed:
fixup! Add a stability policy to the release strategy

mattcaswell commented:
Updates pushed addressing latest feedback. BTW the latest changes imply that we cannot make any more structures opaque until we have documented them as deprecated for 5 years. Do we want to say that?

t-j-h commented:
No. We don't want to be boxed in to leaving a pile of things we planned to sort out and didn't get get to for a much longer time period.

mattcaswell committed:
fixup! Add a stability policy to the release strategy

mattcaswell commented:
I added an exception for making structures opaque - so the first point now says:

"No existing public interface can be modified except where changes are unlikely to break source compatibility or where structures are made opaque."

levitte reviewed (approved):
I'm happy with this

mlindner commented on policies/releasestrat.html:
```diff
@@ -20,27 +20,61 @@ <h5>
 	  <ul>
-	    <li>Version 1.1.0 will be supported until one year after the release of 1.1.1</li>
+	    <li>The next version of OpenSSL will be 3.0.0.</li>
+	    <li>Version 1.1.1 will be supported until 2023-09-11 (LTS).</li>
```
Because it's an LTS release, I would have thought 1.1.1 would be supported until the end of the terminating year (end of 2023), is that not the case?

mattcaswell replied:
No. 1.1.1 is supported for 5 years from its release date.

kaduk commented on policies/releasestrat.html:
```diff
@@ -55,65 +89,94 @@ <h5>
+	    has been in place in an LTS stable release. The original interface must
+	    also have been documented as deprecated for at least 5 years. A public
+	    interface is any function, structure or macro declared in a public
+        header file.</li>
```
indentation nit?

mattcaswell replied:
Fixed.

mattcaswell committed:
fixup! Add a stability policy to the release strategy

levitte reviewed (approved):
Still happy

levitte commented:
BTW, shouldn't this be out of WIP by now?

mattcaswell commented:
> BTW, shouldn't this be out of WIP by now?

Well I was leaving it in WIP until the OMC vote had concluded.

levitte commented:
Ah ok

mattcaswell commented:
Taken out of WIP

mattcaswell committed:
Update the release strategy modification date

mattcaswell commented:
This has now been approved by the OMC. I added an additional commit to update the release date. Please can I get a reconfirmation on this?

levitte reviewed (approved):
Yup

mattcaswell commented:
Pushed. Thanks.
